Add reviews

stepanzl opened the pull request:
Реализована функциональность «Отзывы».

stepanzl committed:
feat: Реализовать систему отзывов с рейтингом полезности

Добавить базовую функциональность отзывов:
1) создание, обновление, удаление и получение отзывов по id и filmId
2) сортировка по рейтингу полезности (useful DESC)
3) возможность ставить лайк и дизлайк отзыву
4) лайк увеличивает рейтинг на +1, дизлайк уменьшает рейтинг на −1
5) переключение лайк/дизлайк корректно пересчитывает рейтинг; повторная реакция игнорируется

Добавить модель Review и DTO ReviewCreateDto, ReviewUpdateDto; контроллер ReviewController; сервис ReviewService; хранилище ReviewDbStorage с таблицами review и review_feedback. Обновить классы валидации и исключений.

stepanzl committed:
test: добавить интеграционные тесты для ReviewService

Создать класс ReviewServiceIntegrationTest для проверки:
1) создания, обновления и удаления отзывов;
2) изменения рейтинга полезности при лайках/дизлайках;
3) корректной сортировки отзывов по useful (DESC);
4) каскадного удаления отзывов с реакциями;
5) обработки ошибок при несуществующих id.

stepanzl committed:
style: Cleanup

h0ttab commented on src/main/java/ru/yandex/practicum/filmorate/model/dto/review/ReviewCreateDto.java:
```diff
@@ -0,0 +1,22 @@
+    @NotNull
+    private Boolean isPositive;
+
+    @NotNull
```
Сюда и в аналогичных случаях можно ещё добавить @Positive , чтобы сразу откидывать отрицательные значения ID.

Это немного избыточно для нашего проекта, но мне кажется, что хуже не будет.

h0ttab commented on src/main/java/ru/yandex/practicum/filmorate/model/dto/review/ReviewUpdateDto.java:
```diff
@@ -0,0 +1,22 @@
+    private Optional<String> content = Optional.empty();
+
+    @Builder.Default
+    private Optional<Boolean> isPositive = Optional.empty();
```
Думаю, что лучше не использовать Optional как тип данных в полях класса. Можно в маппере отзывов оборачивать значения в Optional.ofNullable() и через resolveOptional() их чекать. Так логика окажется собрана в одном месте, а в DTO будут только чистые типы данных.

stepanzl replied:
Тоже так думал когда писал, но на тот момент решил подстроиться под имеющийся код) 
Правку внёс

h0ttab commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDbStorage.java:
```diff
@@ -0,0 +1,186 @@
+                WHERE id = ?;
+                """;
+        jdbcTemplate.update(query, review.getContent(), review.getIsPositive(), review.getReviewId());
+        return findById(review.getReviewId());
```
Тут тоже можно вернуть весь объект review, даже ничего проставлять не надо. Если jdbcTemplate не выплюнул ошибку, значит всё обновилось и объект в базе лежит такой же. Лучше в тестах провернуть такую двойную проверку, чтобы точно знать что review на входе в метод совпадает с объектом из БД после обновления.

h0ttab commented on src/main/resources/schema.sql:
```diff
@@ -53,4 +55,23 @@ CREATE TABLE IF NOT EXISTS friends (
+
+CREATE TABLE IF NOT EXISTS review (
+    id SERIAL PRIMARY KEY,
+    content VARCHAR(1000) NOT NULL,
```
Можно использовать тип данных TEXT - у нас это поле не индексируется, и ограничений по длине нет (можно задать в коде, например 5 тыс. символов). Если говорить о реальной жизни, то отзывы бывают большими - такие киноманы порой пишут, что там и 3-4 тысячи символов может быть.

h0ttab reviewed (approved):
Явных проблем не обнаружил, функционал работает корректно.

В комментариях оставил некоторые предложения по улучшению.

stepanzl committed:
refactor(dto): убрать Optional из ReviewUpdateDto и перенести обработку в маппер

Убран Optional из полей content и isPositive в ReviewUpdateDto.
Логика оборачивания значений в Optional.ofNullable() перенесена в ReviewMapper.
Обновлён тест updateShouldModifyProvidedFields под новый контракт DTO.

stepanzl committed:
refactor(db): изменить тип review.content на TEXT и убрать лишние SELECT

Поле content в таблице review переведено на TEXT.
В create() и update() удалены дополнительные вызовы findById().
Добавлена проверка совпадения данных в тесте updateShouldModifyProvidedFields.

stepanzl committed:
refactor(storage): упростить логику добавления и удаления реакций

applyReaction разбита на небольшие методы (tryFlipReaction, tryInsertReaction, updateUseful) для повышения читаемости и соблюдения DRY, SRP.
removeUseful и removeUseless приведены к единому стилю через общий приватный метод removeReaction.

stepanzl commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDbStorage.java:
```diff
@@ -0,0 +1,186 @@
+        if (deleted > 0) {
+            jdbcTemplate.update("UPDATE review SET useful = useful + 1 WHERE id = ?;", reviewId);
+        }
+    }
```
И правда, тут я почему то решил на двух стульях посидеть, applyReaction сделал, а remove не стал.. 
Переделал это + applyReaction заодно переписал почитаемее, а то глазам тяжело смотреть на него
